libass/ass_render.c

```diff
@@ -2140,6 +2141,50 @@ static void calculate_rotation_params(ASS_Renderer *render_priv, ASS_DRect *bbox
 }
 
 
+static int quantize_blur(double radius, int32_t *shadow_mask)
+{
+    // Gaussian filter kernel (1D):
+    // G(x, r2) = exp(-x^2 / (2 * r2)) / sqrt(2 * pi * r2),
+    // r = 64 * radius, r2 = r^2.
+
+    // Difference between kernels with different but near r2:
+    // G(x, r2 + dr2) - G(x, r2) ~= dr2 * G(x, r2) * (x^2 - r2) / (2 * r2^2).
+    // Maximal possible error relative to full pixel value
+    // is half of integral of absolute value of that difference.
+    // E_max ~= dr2 / 2 * integral(G(x, r2) * |x^2 - r2| / (2 * r2^2), x)
+    //  = dr2 / (4 * r2) * integral(G(y, 1) * |y^2 - 1|, y)
+    //  = dr2 / (4 * r2) * 4 / sqrt(2 * pi * e)
+    //  ~ dr2 / (4 * r2) ~= dr / (2 * r).
+    // E_max ~ BLUR_ACCURACY / 2 as we have 2 dimensions.
+
+    // To get discretized blur radius solve the following
+    // differential equation (n--quantization index):
+    // dr(n) / dn = BLUR_ACCURACY * r + TRANSFORM_ACCURACY, r(0) = 0,
+    // r(n) = (exp(BLUR_ACCURACY * n) - 1) * TRANSFORM_ACCURACY / BLUR_ACCURACY,
+    // n = log(1 + r * BLUR_ACCURACY / TRANSFORM_ACCURACY) / BLUR_ACCURACY.
+
+    // To get shadow offset quantization estimate difference of
+    // G(x + dx, r2) - G(x, r2) ~= dx * G(x, r2) * (-x / r2).
+    // E_max ~= dx / 2 * integral(G(x, r2) * |x| / r2), x)
```

Comment:
I dislike adding new argument for `G(x, r2)` but renaming `dx` indeed makes sense. As for replacing finite difference with derivative: yes indeed in this case it's actually upper bound.

Reply:
On second thought I decided against renaming `dx`: in previous formula I have `r2 + dr2` and here `x + dx` so renaming `dx` alone make that inconsistent. Replacing `x` with `x - x0` and `dx` with `dx0` only adds unnecessary complication. Also my integral notation do not have `dx` so there is no collision here. Maybe better to add note somewhere that by `d...` I mean finite difference (&Delta;...) and not differential.

Reply:
You already say that in other comments; but actually you do mean the differential.

The problem with this line is that you integrate over `x` but factor out `dx` as a constant. Regardless of whether you do explicitly write `dx` in the integral, this does not make sense at face value—like I said, it’s an abuse of notation, which mathematicians do all the time, but it would be nice to avoid it.

Reply:
The only place where I use differentials is `dr(n) / dn = ...`, in other places should be finite differences. I'm using differential equalities to approximate finite differences but only with `~=` sign.

As for notation, `x` and `dx` are two independent terms, so there shouldn't be a problem with correctness. They are related terms but that relation is different than most obvious (due to deficiency of ASCII notation). Of course, by `x` here i mean something like `x - x0` but I think that twofold expansion of expressions hurt understanding more than slightly nonstandard notation.